File: examples/peg.js
load('lib/canopy.js');
load('examples/benchmark.js');
load('examples/pegjs/peg.js');

grammar = 'grammar CompiledPEG\
  \
  grammar             <- space* grammar_name\
                         rules:(space* grammar_rule)+ space*\
  \
  grammar_name        <- "grammar " object_identifier\
  \
  grammar_rule        <- identifier assignment parsing_expression\
  \
  assignment          <- space+ "<-" space+\
  \
  parsing_expression  <- choice_expression / choice_part\
  \
  parenthesised_expression <- "(" space* parsing_expression space* ")"\
  \
  choice_expression   <- first_expression:choice_part\
                         rest_expressions:(space+ "/" space+ expression:choice_part)+\
  \
  choice_part         <- (sequence_expression / sequence_part)\
                         (space+ type_expression)?\
  \
  type_expression     <- "<" object_identifier ">"\
  \
  sequence_expression <- first_expression:sequence_part\
                         rest_expressions:(space+ sequence_part)+\
  \
  sequence_part       <- label? expression:(quantified_atom / atom)\
  \
  quantified_atom     <- atom quantifier\
  \
  atom                <- parenthesised_expression\
                       / predicated_atom\
                       / reference_expression\
                       / string_expression\
                       / any_char_expression\
                       / char_class_expression\
  \
  predicated_atom     <- predicate:("&" / "!") atom\
  \
  reference_expression <- identifier !assignment\
  \
  string_expression   <- "\\"" ("\\\\" . / [^"])* "\\""\
  \
  any_char_expression <- "."\
  \
  char_class_expression <- "[" "^"? ("\\\\" . / [^\\]])+ "]"\
  \
  label               <- identifier ":"\
  \
  object_identifier   <- identifier ("." identifier)*\
  \
  identifier          <- [a-zA-Z_$] [a-zA-Z0-9_$]*\
  \
  quantifier          <- "?" / "*" / "+"\
  \
  space               <- [\\s\\n\\r\\t]';

Canopy.compile(grammar);

benchmark('Canopy parser', 20, function() {
  CompiledPEGParser.parse(grammar);
});

benchmark('PEG.js parser', 20, function() {
  PEGParser.parse(grammar);
});

File: examples/parslet/peg.rb
class ParsletPEG < Parslet::Parser
  root(:grammar)

  rule(:grammar)                  { __.repeat(0) >> grammar_name >>
                                    (__.repeat(0) >> grammar_rule).repeat(1).as(:rules) >>
                                    __.repeat(0) }

  rule(:grammar_name)             { str('grammar') >> str(':').maybe >> __.repeat(1) >> object_identifier }

  rule(:grammar_rule)             { identifier >> assignment >> parsing_expression }

  rule(:assignment)               { __.repeat(1) >> str('<-') >> __.repeat(1) }

  rule(:parsing_expression)       { choice_expression | choice_part }

  rule(:parenthesised_expression) { str('(') >> __.repeat(0) >> parsing_expression >>
                                    __.repeat(0) >> str(')') }

  rule(:choice_expression)        { choice_part.as(:first_part) >>
                                    (__.repeat(1) >> str('/') >> __.repeat(1) >> choice_part.as(:expression)).repeat(1).as(:rest) }

  rule(:choice_part)              { (action_expression | sequence_expression | sequence_part) >>
                                    (__.repeat(1) >> type_tag).maybe }

  rule(:action_expression)        { actionable_expression >> __.repeat(1) >> action_tag }

  rule(:actionable_expression)    { str('(') >> __.repeat(0) >> actionable_expression >> __.repeat(0) >> str(')') |
                                    sequence_expression |
                                    repeated_atom |
                                    terminal_node }

  rule(:action_tag)               { str('%') >> identifier }

  rule(:type_tag)                 { str('<') >> object_identifier >> str('>') }

  rule(:sequence_expression)      { sequence_part.as(:first_part) >>
                                    (__.repeat(1) >> sequence_part.as(:expression)).repeat(1).as(:rest) }

  rule(:sequence_part)            { label.maybe >> (maybe_atom | repeated_atom | atom).as(:expression) }

  rule(:maybe_atom)               { atom >> str('?') }

  rule(:repeated_atom)            { atom >> (str('*') | str('+')).as(:quantifier) }

  rule(:atom)                     { parenthesised_expression |
                                    predicated_atom |
                                    reference_expression |
                                    terminal_node }

  rule(:terminal_node)            { string_expression |
                                    ci_string_expression |
                                    char_class_expression |
                                    any_char_expression }

  rule(:predicated_atom)          { (str('&') | str('!')).as(:predicate) >> atom }

  rule(:reference_expression)     { identifier >> assignment.absent? }

  rule(:string_expression)        { str('"') >> (str('\\') >> any | match('[^"]')).repeat(0) >> str('"') |
                                    str("'") >> (str('\\') >> any | match("[^']")).repeat(0) >> str("'") }

  rule(:ci_string_expression)     { str('`') >> (str('\\') >> any | match('[^`]')).repeat(0) >> str('`') }

  rule(:any_char_expression)      { str('.') }

  rule(:char_class_expression)    { str('[') >> str('^').maybe >> (str('\\') >> any | match('[^\\]]')).repeat(1) >> str(']') }

  rule(:label)                    { identifier >> str(':') }

  rule(:object_identifier)        { identifier >> (str('.') >> identifier).repeat(0) }

  rule(:identifier)               { match('[a-zA-Z_]') >> match('[a-zA-Z0-9_]').repeat(0) }

  rule(:__)                       { match('[\\s]') | comment }

  rule(:comment)                  { str('#') >> match('[^\n]').repeat(0) }
end
